<?xml version="1.0" ?>
<!-- =================================================================================== -->
<!-- |    This document was autogenerated by xacro from urdf/carModel.xacro            | -->
<!-- |    EDITING THIS FILE BY HAND IS NOT RECOMMENDED                                 | -->
<!-- =================================================================================== -->
<robot name="car_model" xmlns:xacro="http://www.ros.org/wiki/xacro">
  <material name="black">
    <color rgba="0 0 0 1"/>
  </material>
  <material name="blue">
    <color rgba="0 0 0.8 1"/>
  </material>
  <material name="white">
    <color rgba="1 1 1 1"/>
  </material>
  <material name="gray">
    <color rgba="0.3 0.3 0.3 1"/>
  </material>
  <link name="base_footprint">
    <visual>
      <geometry>
        <box size="0.001 0.001 0.001"/>
      </geometry>
      <origin rpy="0 0 0" xyz="0 0 0"/>
    </visual>
    <collision>
      <geometry>
        <box size="0.001 0.001 0.001"/>
      </geometry>
      <origin rpy="0 0 0" xyz="0 0 0"/>
    </collision>
    <inertial>
      <mass value="0.0001"/>
      <inertia ixx="1.0" ixy="0.0" ixz="0.0" iyy="1.0" iyz="0.0" izz="1.0"/>
    </inertial>
  </link>
  <gazebo reference="base_footprint">
    <material>Gazebo/Green</material>
    <turnGravityOff>false</turnGravityOff>
  </gazebo>
  <link name="base_link">
    <visual>
      <geometry>
        <box size="0.25 0.5 0.0625"/>
      </geometry>
      <material name="black"/>
      <origin rpy="0 0 0" xyz="0 0 0.125"/>
    </visual>
    <collision>
      <geometry>
        <box size="0.25 0.5 0.0625"/>
      </geometry>
      <origin rpy="0 0 0" xyz="0 0 0.125"/>
    </collision>
    <inertial>
      <mass value="5"/>
      <inertia ixx="1.0" ixy="0.0" ixz="0.0" iyy="1.0" iyz="0.0" izz="1.0"/>
    </inertial>
  </link>
  <joint name="base_footprint_joint" type="fixed">
    <axis xyz="0 0 1"/>
    <origin xyz="0 0 0"/>
    <parent link="base_footprint"/>
    <child link="base_link"/>
  </joint>
  <gazebo reference="base_link">
    <material>Gazebo/Black</material>
  </gazebo>
  <link name="camera">
    <visual>
      <geometry>
        <box size="0.1 0.1 0.1"/>
      </geometry>
      <material name="blue"/>
    </visual>
    <collision>
      <geometry>
        <box size="0.1 0.1 0.1"/>
      </geometry>
    </collision>
    <inertial>
      <mass value="1e-05"/>
      <inertia ixx="1.0" ixy="0.0" ixz="0.0" iyy="1.0" iyz="0.0" izz="1.0"/>
    </inertial>
  </link>
  <joint name="camera_joint" type="fixed">
    <axis xyz="0 0 1"/>
    <origin rpy="0 0 0" xyz="0 0.3125 0.1875"/>
    <parent link="base_footprint"/>
    <child link="camera"/>
  </joint>
  <gazebo reference="camera">
    <!--from http://gazebosim.org/tutorials?tut=ros_gzplugins-->
    <sensor name="camera" type="camera">
      <update_rate>33.3</update_rate>
      <camera name="head">
        <horizontal_fov>1.3962634</horizontal_fov>
        <image>
          <width>&quot;640&quot;</width>
          <height>&quot;480&quot;</height>
          <format>R8G8B8</format>
        </image>
        <clip>
          <near>0.02</near>
          <far>300</far>
        </clip>
        <noise>
          <type>gaussian</type>
          <!-- Noise is sampled independently per pixel on each frame.
                                                That pixel's noise value is added to each of its color
                                                channels, which at that point lie in the range [0,1]. -->
          <mean>0.0</mean>
          <stddev>0.007</stddev>
        </noise>
      </camera>
      <plugin filename="libgazebo_ros_camera.so" name="camera_controller">
        <alwaysOn>true</alwaysOn>
        <updateRate>0.0</updateRate>
        <cameraName>robot/camera</cameraName>
        <imageTopicName>image_raw</imageTopicName>
        <cameraInfoTopicName>camera_info</cameraInfoTopicName>
        <frameName>camera_link</frameName>
        <hackBaseline>0.07</hackBaseline>
        <distortionK1>0.0</distortionK1>
        <distortionK2>0.0</distortionK2>
        <distortionK3>0.0</distortionK3>
        <distortionT1>0.0</distortionT1>
        <distortionT2>0.0</distortionT2>
      </plugin>
    </sensor>
  </gazebo>
  <link name="front_axle">
    <visual>
      <geometry>
        <cylinder length="0.4" radius="0.0625"/>
      </geometry>
      <material name="gray"/>
    </visual>
    <collision>
      <geometry>
        <cylinder length="0.4" radius="0.0625"/>
      </geometry>
    </collision>
    <inertial>
      <mass value="3"/>
      <inertia ixx="1.0" ixy="0.0" ixz="0.0" iyy="1.0" iyz="0.0" izz="1.0"/>
    </inertial>
  </link>
  <joint name="front_axle_to_base" type="fixed">
    <axis xyz="1 0 0"/>
    <limit effort="1000.0" lower="-0.785" upper="0.785" velocity="1.0"/>
    <mimic joint="base_footprint_joint" multiplier="1"/>
    <parent link="base_footprint"/>
    <child link="front_axle"/>
    <origin rpy="0 -1.57 0" xyz="0 0.3 0.125"/>
  </joint>
  <gazebo reference="front_axle">
    <material>Gazebo/Grey</material>
  </gazebo>
  <link name="back_axle">
    <visual>
      <geometry>
        <cylinder length="0.4" radius="0.0625"/>
      </geometry>
      <material name="gray"/>
    </visual>
    <collision>
      <geometry>
        <cylinder length="0.4" radius="0.0625"/>
      </geometry>
    </collision>
    <inertial>
      <mass value="3"/>
      <inertia ixx="1.0" ixy="0.0" ixz="0.0" iyy="1.0" iyz="0.0" izz="1.0"/>
    </inertial>
  </link>
  <joint name="back_axle_to_base" type="fixed">
    <axis xyz="1 0 0"/>
    <limit effort="1000.0" lower="-0.785" upper="0.785" velocity="1.0"/>
    <mimic joint="base_footprint_joint" multiplier="1"/>
    <parent link="base_footprint"/>
    <child link="back_axle"/>
    <origin rpy="0 -1.57 0" xyz="0 -0.3 0.125"/>
  </joint>
  <gazebo reference="back_axle">
    <material>Gazebo/Grey</material>
  </gazebo>
  <link name="front_left_wheel">
    <visual>
      <geometry>
        <cylinder length="0.125" radius="0.125"/>
      </geometry>
      <material name="white"/>
    </visual>
    <collision>
      <geometry>
        <cylinder length="0.125" radius="0.125"/>
      </geometry>
    </collision>
    <inertial>
      <mass value="20"/>
      <inertia ixx="1.0" ixy="0.0" ixz="0.0" iyy="1.0" iyz="0.0" izz="1.0"/>
    </inertial>
  </link>
  <joint name="front_left_wheel_to_front_axle" type="revolute">
    <axis xyz="1 0 0"/>
    <limit effort="1000.0" lower="-0.393" upper="0.785" velocity="1.0"/>
    <parent link="base_footprint"/>
    <child link="front_left_wheel"/>
    <origin rpy="0 -1.57 0" xyz="-0.25 0.3 0.125"/>
  </joint>
  <gazebo reference="front_left_wheel">
    <material>Gazebo/White</material>
  </gazebo>
  <link name="front_right_wheel">
    <visual>
      <geometry>
        <cylinder length="0.125" radius="0.125"/>
      </geometry>
      <material name="white"/>
    </visual>
    <collision>
      <geometry>
        <cylinder length="0.125" radius="0.125"/>
      </geometry>
    </collision>
    <inertial>
      <mass value="20"/>
      <inertia ixx="1.0" ixy="0.0" ixz="0.0" iyy="1.0" iyz="0.0" izz="1.0"/>
    </inertial>
  </link>
  <joint name="front_right_wheel_to_front_axle" type="revolute">
    <axis xyz="1 0 0"/>
    <limit effort="1000.0" lower="-0.785" upper="0.393" velocity="1.0"/>
    <parent link="base_footprint"/>
    <child link="front_right_wheel"/>
    <origin rpy="0 -1.57 0" xyz="0.25 0.3 0.125"/>
  </joint>
  <gazebo reference="front_right_wheel">
    <material>Gazebo/White</material>
  </gazebo>
  <link name="back_left_wheel">
    <visual>
      <geometry>
        <cylinder length="0.125" radius="0.125"/>
      </geometry>
      <material name="white"/>
    </visual>
    <collision>
      <geometry>
        <cylinder length="0.125" radius="0.125"/>
      </geometry>
    </collision>
    <inertial>
      <mass value="20"/>
      <inertia ixx="1.0" ixy="0.0" ixz="0.0" iyy="1.0" iyz="0.0" izz="1.0"/>
    </inertial>
  </link>
  <joint name="back_left_wheel_to_back_axle" type="revolute">
    <axis xyz="0 0 1"/>
    <limit effort="1000.0" lower="-3140.0" upper="3140.0" velocity="1.0"/>
    <parent link="base_footprint"/>
    <child link="back_left_wheel"/>
    <origin rpy="0 -1.57 0" xyz="-0.25 -0.3 0.125"/>
  </joint>
  <gazebo reference="back_left_wheel">
    <material>Gazebo/White</material>
  </gazebo>
  <link name="back_right_wheel">
    <visual>
      <geometry>
        <cylinder length="0.125" radius="0.125"/>
      </geometry>
      <material name="white"/>
    </visual>
    <collision>
      <geometry>
        <cylinder length="0.125" radius="0.125"/>
      </geometry>
    </collision>
    <inertial>
      <mass value="20"/>
      <inertia ixx="1.0" ixy="0.0" ixz="0.0" iyy="1.0" iyz="0.0" izz="1.0"/>
    </inertial>
  </link>
  <joint name="back_right_wheel_to_back_axle" type="revolute">
    <axis xyz="0 0 1"/>
    <limit effort="1000.0" lower="-3140.0" upper="3140.0" velocity="1.0"/>
    <parent link="base_footprint"/>
    <child link="back_right_wheel"/>
    <origin rpy="0 -1.57 0" xyz="0.25 -0.3 0.125"/>
  </joint>
  <gazebo reference="back_right_wheel">
    <material>Gazebo/White</material>
  </gazebo>
  <!--
-->
  <gazebo>
    <plugin filename="libgazebo_ros_diff_drive.so" name="differential_drive_controller">
      <alwaysOn>true</alwaysOn>
      <commandTopic>cmd_vel</commandTopic>
      <leftJoint>front_left_wheel_to_front_axle</leftJoint>
      <legacyMode>false</legacyMode>
      <odometryFrame>odom</odometryFrame>
      <odometryTopic>odom</odometryTopic>
      <publishTf>100</publishTf>
      <publishWheelJointState>false</publishWheelJointState>
      <publishWheelTF>true</publishWheelTF>
      <rightJoint>front_right_wheel_to_front_axle</rightJoint>
      <robotBaseFrame>base_link</robotBaseFrame>
      <robotNamespace>robot::base_footprint</robotNamespace>
      <rosDebugLevel>na</rosDebugLevel>
      <wheelTorque>5</wheelTorque>
      <updateRate>0.1</updateRate>
      <wheelAcceleration>2</wheelAcceleration>
      <wheelDiameter>0.25</wheelDiameter>
      <wheelSeparation>0.4</wheelSeparation>
    </plugin>
  </gazebo>
  <!--
        <leftJoint>front_right_wheel_to_front_axle</leftJoint>
        <rightJoint>front_left_wheel_to_front_axle</rightJoint>
        <leftJoint>back_right_wheel_to_back_axle</leftJoint>
        <rightJoint>back_left_wheel_to_back_axle</rightJoint>
        -->
</robot>

